File: src/main/java/org/arbor/gtnn/api/recipe/NeutronActivatorCondition.java
package org.arbor.gtnn.api.recipe;

import com.google.gson.JsonObject;
import com.gregtechceu.gtceu.api.machine.MetaMachine;
import com.gregtechceu.gtceu.api.machine.trait.RecipeLogic;
import com.gregtechceu.gtceu.api.recipe.GTRecipe;
import com.gregtechceu.gtceu.api.recipe.RecipeCondition;
import lombok.Getter;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.chat.Component;
import net.minecraft.util.GsonHelper;
import org.arbor.gtnn.api.machine.multiblock.NeutronActivator;
import org.jetbrains.annotations.NotNull;

@Getter
public class NeutronActivatorCondition extends RecipeCondition {
    public static final NeutronActivatorCondition INSTANCE = new NeutronActivatorCondition();
    private int evRange = 0;
    public NeutronActivatorCondition(int max, int min){
        super();
        this.evRange = max * 10000 + min;
    }

    public NeutronActivatorCondition() {
        super();
    }

    @Override
    public String getType() {
        return "neutron_activator_condition";
    }

    @Override
    public Component getTooltips() {
        int max = evRange % 10000;
        int min = evRange / 10000;
        return Component.translatable("gtnn.recipe.condition.neutron_activator_condition_tooltip", max, min);
    }

    @Override
    public boolean test(@NotNull GTRecipe gtRecipe, @NotNull RecipeLogic recipeLogic) {
        return NeutronActivator.checkNeutronActivatorCondition((MetaMachine) recipeLogic.machine, gtRecipe);
    }

    @Override
    public RecipeCondition createTemplate() {
        return new NeutronActivatorCondition();
    }

    @NotNull
    @Override
    public JsonObject serialize() {
        JsonObject value = super.serialize();
        value.addProperty("evRange", this.evRange);
        return value;
    }

    @Override
    public RecipeCondition deserialize(@NotNull JsonObject config) {
        super.deserialize(config);
        this.evRange = GsonHelper.getAsInt(config, "evRange", 0);
        return this;
    }

    @Override
    public void toNetwork(FriendlyByteBuf buf) {
        super.toNetwork(buf);
        buf.writeInt(this.evRange);
    }

    @Override
    public RecipeCondition fromNetwork(FriendlyByteBuf buf) {
        super.fromNetwork(buf);
        this.evRange = buf.readInt();
        return this;
    }
}

File: src/main/kotlin/dev/arbor/gtnn/common/recipe/NeutronActivatorCondition.kt
package dev.arbor.gtnn.common.recipe

import com.gregtechceu.gtceu.api.machine.MetaMachine
import com.gregtechceu.gtceu.api.machine.trait.RecipeLogic
import com.gregtechceu.gtceu.api.recipe.GTRecipe
import com.gregtechceu.gtceu.api.recipe.RecipeCondition
import com.gregtechceu.gtceu.api.recipe.condition.RecipeConditionType
import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import dev.arbor.gtnn.common.machine.multiblock.NeutronActivatorMachine.Companion.checkNeutronActivatorCondition
import dev.arbor.gtnn.data.GTNNRecipeConditions
import net.minecraft.network.chat.Component

class NeutronActivatorCondition(max: Int, min: Int) : RecipeCondition<NeutronActivatorCondition>() {
    companion object {
        val CODEC: Codec<NeutronActivatorCondition> = RecordCodecBuilder
            .create { instance: RecordCodecBuilder.Instance<NeutronActivatorCondition> ->
                isReverse(instance)
                    .and(Codec.INT.fieldOf("evRange").forGetter { it.evRange })
                    .apply(instance, ::NeutronActivatorCondition)
            }
    }

    var evRange: Int = 0

    constructor() : this(0, 0)

    constructor(evRange: Int) : this() {
        this.evRange = evRange
    }

    constructor(isReverse: Boolean, evRange: Int) : this(evRange) {
        super.isReverse = isReverse
    }

    init {
        this.evRange = max * 10000 + min
    }

    override fun getType(): RecipeConditionType<NeutronActivatorCondition> {
        return GTNNRecipeConditions.NEUTRON_ACTIVATOR
    }

    override fun getTooltips(): Component {
        val max = evRange % 10000
        val min = evRange / 10000
        return Component.translatable("gtnn.recipe.condition.neutron_activator_condition_tooltip", max, min)
    }

    override fun testCondition(gtRecipe: GTRecipe, recipeLogic: RecipeLogic): Boolean {
        return checkNeutronActivatorCondition((recipeLogic.machine as MetaMachine), gtRecipe)
    }

    override fun createTemplate(): NeutronActivatorCondition {
        return NeutronActivatorCondition()
    }
}
